keps/sig-multicluster/4322-cluster-inventory/README.md

```diff
@@ -0,0 +1,1012 @@
+manager must set a different value of `spec.clusterManager.name` when
+the cluster is created.
+
+### Spec
+
+#### Display name
+
+It is a human-readable name of the cluster set by the consumer of the
+cluster.
+
+#### Cluster Manager
+
+An immutable field set by a cluster manager when this cluster resources
+is created by it. Each cluster manager instance should set a different
+values to this field.
+
+In addition, a predefined label with key "x-k8s.io/cluster-manager" needs
+to be added by the cluster manager upon creation. The value of the label
+MUST be the same as the name of the cluster manager. The purpose of this
+label is to make filter clusters from different cluster managers easier.
+
+### Status
+
+#### Version
+
+Kubernetes version of the cluster
+
+Versions of the kubernetes can let consumers understand the capability
+of the kubernetes, such as what API is supported.
+
+With recent conversations about kube-apiserver and enabled featureset version,
+it is possible to incorporate other version relating to the cluster, such as
+minimum kubelet version, maximum kubelet version, and enabled featureset version.
+
+#### Properties
+
+Name/value pairs to represent properties of the clusters. It could be a
+collection of ClusterProperty resources, but could also be info based on
+other implementations. The name of the cluster property can be predefined
+name from ClusterProperty resources and is allowed to be customized by
+different cluster managers.
+
+#### Conditions
+
+Record cluster’s healthiness condition and easy to extend, conforming to
+metav1.Condition format.
+
+Predefined condition types:
+- Healthy conditions indicate the cluster is in a good state. (state:
+  True/False/Unknown). Healthiness can have different meanings in
+  different scenarios. We will have multiple condition types to define
+  healthiness:
+  - ControlPlaneHealthy is to define whether the controlplane of the
+    cluster is in the healthy state.
+    * apiserver/healthz
+    * controller-manager/healthz
+    * scheduler/healthz
+    * etcd/healthz
+  - <<[UNRESOLVED]>> AllNodesHealthy is to define if the nodes in the cluster are in a
+    healthy state. If one node is not healthy, the status of NodeHealthy
+    is set to false with the message indicating details. (todo
+    tolerance, it should be configurable)
+  It would be useful to collect the healthiness of other subsystems in
+  the cluster, e.g. network, dns, storage, or ingress. However, it is
+  not easy to collect that information in a common way with different
+  implementations of network or storage providers. We decide not to
+  include other subsystems healthiness conditions in the initial phase.
+- Joined: indicate the cluster is under management by the cluster manager.
+  The status of the cluster SHOULD be updated by the cluster manager under
+  this condition.
+
+## API Example
+
+```yaml
+apiVersion: multicluster.x-k8s.io/v1alpha1
+kind: InventoryCluster
```

Comment:
Do we have a mention why we aren't able to re-use an existing standard API to represent a cluster. For example:

- https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/cluster_types.go#L489-L495

Reply:
Right. Maybe we can add a mention to why consumers and providers prefer a more "neutral" API like this one over something like clusterapi which imo has a coupling with cluster provisioning.

Reply:
To be fair we are fully qualifying the resource under the `multicluster.x-k8s.io/v1alpha1` group/version in this proposal, whereas CAPI is represented under `cluster.x-k8s.io/v1beta1`. So I don't think we're making any serious violation.

Sorry to be annoying but I'd probably place a vote for `ClusterRecord` instead of `InventoryCluster` or `ClusterInventory`. Or is there a better word than "record" to describe a unit in an inventory?

Reply:
I am not a big fan of `ClusterRecord` or `ClusterProfile` or something similar if we are planning to have the API access info in there as well for phase 2. WDYT?

Reply:
I think something like `ClusterProfile` or `ClusterRecord` would actually make a lot of sense for phase 2 needs too - semantically a "profile" and a "record" are both summary descriptions of an entity (the cluster in this case), which would both be written to by a user as a means of describing the entity (in Kubernetes APIs this would be through the `spec` section) or updated/categorized/reported on by a system or administrator (this would be a controller updating the `status` section).

I'll digress a bit about into a pedantic critique of specifically why I think the `InventoryCluster` term feels awkward - in many other Kubernetes APIs, the terminology pattern for combined camel-case APIs is for the last half of the term to be a generic "type" of resource, and the first half to be a means of qualifying the scope or role of the resource - examples include `DaemonSet`, `ReplicaSet`, `StatefulSet` and `RoleBinding`, `ClusterRoleBinding`. Following this pattern, the structure of the term `InventoryCluster` could imply that the object it represents is a cluster used exclusively for inventory purposes (think how https://open-cluster-management.io/concepts/architecture/#hub-spoke-architecture uses terms structured this way for "Hub Cluster" and "Managed Cluster"). Additionally, cluster-scoped resources typically follow the above-described pattern by using a `Cluster*` _prefixed_ term to indicate their scope (see `ClusterRoleBinding`, [`ClusterProperty`](https://github.com/kubernetes-sigs/about-api), etc), and I believe it would help users better understand the usage of this API if it followed terminology patterns more similar to existing APIs.

Reply:
Right. We should follow convention and not use something that feels unnatural.
Let's gather some more feedback and hopefully come to a conclusion in the next SIG MC meeting.

Reply:
@jackfrancis 

I think this API serves a different purpose from cluster-api. cluster-api describes how to construct a cluster. This API describes the cluster for whoever wants to use the cluster in a multi-cluster environment.

We should allow customers who don't use cluster API to provision their cluster to be able to describe a cluster in a multi-cluster environment.

Reply:
The `Cluster` interface in Cluster API just _decribes a cluster_, similar to what is being designed here. I agree that the way that Cluster API describes a cluster (ref: https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/cluster_types.go#L45) is probably not general enough for re-use in other related concerns (which is unfortunate), including this one.

And certainly agree that we don't want to require Cluster API for the inventory use case, just wondering if we should add a note for objective observers who might wonder (with justification IMO!) "Why does the Kubernetes project define a Cluster many different ways?"

Reply:
> The `Cluster` in Cluster API just _decribes a cluster_, similar to what is being designed here. I agree that the way that Cluster API describes a cluster (ref: https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/cluster_types.go#L45) is probably not general enough for re-use in other related concerns (which is unfortunate), including this one.

The Cluster API `Cluster` is more than that, isn't it? It describes the state of the cluster as the operator wants it to be. InventorCluster (or whatever we go with :) ) is closer to Cluster API's `Cluster.status`. To me this shows a mismatch - beyond just the fact that Cluster API shouldn't be required - the two resources are owned by different entities. Cluster API's `Cluster` is owned by the platform administrator provisioning clusters, the sig-mc `ClusterInventory` is owned by the cluster manager that created the cluster it represents.

Reply:
> It seems that this is exactly what we are trying to achieve here. Why is that not an accurate description of the name?

Following the examples above, it could be inferred as describing to the _internal contents_ of the cluster, like "a Kubernetes cluster which only holds inventory records" as opposed to the _description_ of a specific cluster in a list of records.